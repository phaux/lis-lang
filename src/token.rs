//! This module defines types which represent [Token]s.

#[derive(Debug, PartialEq, Clone)]
pub struct Token {
    pub sigil: Sigil,
    /// Byte offset of the token in the input string.
    pub offset: std::ops::Range<usize>,
    /// Position of the token in the input string.
    pub pos: std::ops::Range<Pos>,
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "`{}` at {}", self.sigil, self.pos.start)
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum Sigil {
    Keyword(Keyword),
    Ident { name: String },
    Num { val: f64 },
    Str { val: String },
    Comment { body: String },
    Eq,
    EqEq,
    Bang,
    BangEq,
    Semi,
    Plus,
    PlusPlus,
    Minus,
    Star,
    Slash,
    Dot,
    Colon,
    Comma,
    ParenL,
    ParenR,
    CurlyL,
    CurlyR,
    Invalid,
}

impl std::fmt::Display for Sigil {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Sigil::Keyword(keyword) => write!(f, "{keyword}"),
            Sigil::Ident { name } => write!(f, "{name}"),
            Sigil::Num { val } => write!(f, "{val:?}"),
            Sigil::Str { val } => write!(f, "{val:?}"),
            Sigil::Comment { .. } => write!(f, "comment"),
            Sigil::Eq => write!(f, "="),
            Sigil::EqEq => write!(f, "=="),
            Sigil::Bang => write!(f, "!"),
            Sigil::BangEq => write!(f, "!="),
            Sigil::Semi => write!(f, ";"),
            Sigil::Plus => write!(f, "+"),
            Sigil::PlusPlus => write!(f, "++"),
            Sigil::Minus => write!(f, "-"),
            Sigil::Star => write!(f, "*"),
            Sigil::Slash => write!(f, "/"),
            Sigil::Dot => write!(f, "."),
            Sigil::Colon => write!(f, ":"),
            Sigil::Comma => write!(f, ","),
            Sigil::ParenL => write!(f, "("),
            Sigil::ParenR => write!(f, ")"),
            Sigil::CurlyL => write!(f, "{{"),
            Sigil::CurlyR => write!(f, "}}"),
            Sigil::Invalid => write!(f, "invalid token"),
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum Keyword {
    True,
    False,
    Nil,
    Let,
    Fn,
    Print,
    If,
    Then,
    Else,
    For,
    While,
    Do,
    Break,
    Continue,
    Return,
    And,
    Or,
}

impl std::str::FromStr for Keyword {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "true" => Ok(Keyword::True),
            "false" => Ok(Keyword::False),
            "nil" => Ok(Keyword::Nil),
            "let" => Ok(Keyword::Let),
            "fn" => Ok(Keyword::Fn),
            "print" => Ok(Keyword::Print),
            "if" => Ok(Keyword::If),
            "then" => Ok(Keyword::Then),
            "else" => Ok(Keyword::Else),
            "for" => Ok(Keyword::For),
            "while" => Ok(Keyword::While),
            "do" => Ok(Keyword::Do),
            "break" => Ok(Keyword::Break),
            "continue" => Ok(Keyword::Continue),
            "return" => Ok(Keyword::Return),
            "and" => Ok(Keyword::And),
            "or" => Ok(Keyword::Or),
            _ => Err(()),
        }
    }
}

impl std::fmt::Display for Keyword {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Keyword::True => write!(f, "true"),
            Keyword::False => write!(f, "false"),
            Keyword::Nil => write!(f, "nil"),
            Keyword::Let => write!(f, "let"),
            Keyword::Fn => write!(f, "fn"),
            Keyword::Print => write!(f, "print"),
            Keyword::If => write!(f, "if"),
            Keyword::Then => write!(f, "then"),
            Keyword::Else => write!(f, "else"),
            Keyword::For => write!(f, "for"),
            Keyword::While => write!(f, "while"),
            Keyword::Do => write!(f, "do"),
            Keyword::Break => write!(f, "break"),
            Keyword::Continue => write!(f, "continue"),
            Keyword::Return => write!(f, "return"),
            Keyword::And => write!(f, "and"),
            Keyword::Or => write!(f, "or"),
        }
    }
}

/// Line and column.
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Default, Clone, Copy)]
pub struct Pos {
    /// 0-based line number.
    pub line: usize,
    /// 0-based column number.
    pub col: usize,
}

impl std::fmt::Display for Pos {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.line, self.col)
    }
}
