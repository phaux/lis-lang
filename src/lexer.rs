//! This module defines the [Lexer].
//! It transforms a string of characters into a stream of [Token]s.

use std::str::FromStr as _;

use crate::token::{Keyword, Pos, Sigil, Token};

pub struct Lexer<'a> {
    input: &'a str,
    /// Current byte offset in the input string.
    offset: usize,
    /// Current line and column.
    pos: Pos,
}

impl<'a> Lexer<'a> {
    #[must_use]
    pub fn new(input: &'a str) -> Self {
        Self {
            input,
            offset: 0,
            pos: Pos::default(),
        }
    }

    fn peek_char(&self) -> Option<char> {
        self.input[self.offset..].chars().next()
    }

    fn next_char(&mut self) -> Option<char> {
        let ch = self.input[self.offset..].chars().next()?;
        if ch == '\n' {
            self.pos.line += 1;
            self.pos.col = 0;
        } else {
            self.pos.col += 1;
        }
        self.offset += ch.len_utf8();
        Some(ch)
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let offset_start = self.offset;
            let pos_start = self.pos;
            let ch = self.next_char()?;
            let sigil = match ch {
                ch if ch.is_whitespace() => continue,
                // Comments: // line and /* block */
                '/' => match self.peek_char() {
                    Some('/') => self.parse_line_comment(),
                    Some('*') => self.parse_block_comment(),
                    _ => Sigil::Slash,
                },
                '=' => match self.peek_char() {
                    Some('=') => {
                        self.next_char();
                        Sigil::EqEq
                    }
                    _ => Sigil::Eq,
                },
                '+' => match self.peek_char() {
                    Some('+') => {
                        self.next_char();
                        Sigil::PlusPlus
                    }
                    _ => Sigil::Plus,
                },
                '!' => match self.peek_char() {
                    Some('=') => {
                        self.next_char();
                        Sigil::BangEq
                    }
                    _ => Sigil::Bang,
                },
                ';' => Sigil::Semi,
                '-' => Sigil::Minus,
                '*' => Sigil::Star,
                '.' => Sigil::Dot,
                ':' => Sigil::Colon,
                ',' => Sigil::Comma,
                '(' => Sigil::ParenL,
                ')' => Sigil::ParenR,
                '{' => Sigil::CurlyL,
                '}' => Sigil::CurlyR,
                '<' => match self.peek_char() {
                    Some('=') => {
                        self.next_char();
                        Sigil::LessEq
                    }
                    _ => Sigil::Less,
                },
                '>' => match self.peek_char() {
                    Some('=') => {
                        self.next_char();
                        Sigil::GreaterEq
                    }
                    _ => Sigil::Greater,
                },
                '"' => self.parse_str(),
                ch if ch.is_alphabetic() || ch == '_' => self.parse_word(ch),
                ch if ch.is_ascii_digit() => self.parse_num(ch),
                _ => Sigil::Invalid,
            };
            return Some(Token {
                sigil,
                offset: offset_start..self.offset,
                pos: pos_start..self.pos,
            });
        }
    }
}

impl Lexer<'_> {
    fn parse_num(&mut self, first: char) -> Sigil {
        let mut val_str = String::new();
        val_str.push(first);
        while let Some(ch) = self.peek_char()
            && (ch.is_ascii_digit() || ch == '.')
        {
            self.next_char();
            val_str.push(ch);
        }
        let Ok(val) = val_str.parse::<f64>() else {
            return Sigil::Invalid;
        };
        Sigil::Num { val }
    }

    fn parse_word(&mut self, first: char) -> Sigil {
        let mut word = String::new();
        word.push(first);
        while let Some(ch) = self.peek_char()
            && (ch.is_alphanumeric() || ch == '_')
        {
            self.next_char();
            word.push(ch);
        }
        match Keyword::from_str(&word) {
            Ok(keyword) => Sigil::Keyword(keyword),
            Err(()) => Sigil::Ident { name: word },
        }
    }

    fn parse_str(&mut self) -> Sigil {
        let mut val = String::new();
        while let Some(ch) = self.next_char() {
            match ch {
                '"' => break,
                '\\' => match self.next_char() {
                    Some('n') => val.push('\n'),
                    Some('t') => val.push('\t'),
                    Some('r') => val.push('\r'),
                    Some('\\') => val.push('\\'),
                    Some('"') => val.push('"'),
                    Some(other) => val.push(other),
                    None => {}
                },
                ch => val.push(ch),
            }
        }
        Sigil::Str { val }
    }

    fn parse_line_comment(&mut self) -> Sigil {
        // Consume the second '/'
        self.next_char();
        let mut body = String::new();
        while let Some(ch) = self.peek_char() {
            if ch == '\n' {
                break;
            }
            self.next_char();
            body.push(ch);
        }
        Sigil::Comment { body }
    }

    fn parse_block_comment(&mut self) -> Sigil {
        // Consume the '*'
        self.next_char();
        let mut body = String::new();
        // Read until closing */ or EOF
        while let Some(ch) = self.next_char() {
            if ch == '*'
                && let Some('/') = self.peek_char()
            {
                self.next_char();
                break;
            }
            body.push(ch);
        }
        Sigil::Comment { body }
    }
}

#[cfg(test)]
#[path = "lexer.tests.rs"]
mod tests;
