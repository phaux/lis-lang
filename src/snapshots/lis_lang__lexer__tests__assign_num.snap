---
source: src/lexer.rs
expression: tokens
---
[
    Token {
        sigil: Ident {
            name: "x",
        },
        range: 1..2,
    },
    Token {
        sigil: Eq,
        range: 3..4,
    },
    Token {
        sigil: Num {
            val: 5.5,
        },
        range: 5..8,
    },
]
